### ISSUE TITLE
the findOne operation always return the first record in the DB even if I passed a undefined value or null value

### ISSUE THREAD

**Issue by alanchenx** [typeorm/typeorm#9408]
### Actual Behavior

```js
// codes below always return the first record in the Database
  console.log("query start")
  const res = await this.userModel.findOne({where: {
    userName: null
  }})
  console.log(res)
  console.log("query end")

// result as below: 
query start
User {
  createdAt: 1661686557000,
  modifiedAt: null,
  deletedAt: null,
  userId: 10,
  pwd: 'xxxxxxx',
  userName: 'xxxx',
  userRole: 0,
  avatar: 'xxxxxxxx',
  config: { theme: 'dark', language: 'zh' }
}
query end
```
### Steps to Reproduce

```typescript
// insert code here
  console.log("query start")
  const res = await this.userModel.findOne({where: {
    userName: null
  }})
  console.log(res)
  console.log("query end")
```

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |     Macos 12.4     |
| Node.js version     | v16.16.0 | 
| Typescript version  | ~4.6.0  | 
| TypeORM version     | ^0.3.10  | 

### Relevant Database Driver(s)

| DB Type           | Reproducible  |
|-------------------| ---           |
| `aurora-mysql`    | no            |
| `aurora-postgres` | no            |
| `better-sqlite3`  | no            |
| `cockroachdb`     | no            |
| `cordova`         | no            |
| `expo`            | no            |
| `mongodb`         | no            |
| `mysql`           | yes            |
| `nativescript`    | no            |
| `oracle`          | no            |
| `postgres`        | no            |
| `react-native`    | no            |
| `sap`             | no            |
| `spanner`         | no            |
| `sqlite`          | no            |
| `sqlite-abstract` | no            |
| `sqljs`           | no            |
| `sqlserver`       | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✅ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by mlandisbqs** [comment]
👀

**Comment by alanchenx** [comment]
@matheusgrandi Yeah thanks. I have done it. But I still want the team to fix this bug. It may cause unexpected problem in production environment.

**Comment by iJhefe** [comment]
change for 
```typescript
import { Equal } from 'typeorm'

// ...

 const res = await this.userModel.findOne({where: {
    userName: Equal(null) // the param on your request
  }})
```

**Comment by cfarvidson** [comment]
I also ran into this one today.

**Comment by gabrielrra** [comment]
Had the same problem

**Comment by mlandisbqs** [comment]
This issue keeps coming up and getting closed. I can't find my original comment but I think the problem is with the implied _contract_ for findOne. The implementation of this function is behaving as if it was named findFirst and not findOne. 

This is actually dangerous since admin users are typically the first record in the db. If you don't catch this you could leak superuser privs to the wrong user. It's happened to me (in a test env caught first) and at least 2 other commenters on this issue every time it gets opened and closed again.

suggestion:
- deprecate findOne so as not to break existing implementations
- deprecate findOneOrFail since it appears to suffers from the same contract confusion
- copy findOne with a new name findFirst which better represents the implied contract
- add a function findExactlyOneOrFail that fails if more than one record matches the criteria - instead of returning the first thing.

Thoughts?

**Comment by cfarvidson** [comment]
I think that @mlandisbqs makes a great point and I like the suggested solution!

**Comment by holuborhee** [comment]
> There's a simple solution for your problem using conditional operator.
> 
> ```js
> const res = userName !== (undefined || null)
> ? await this.userModel.findOne({where: {
>     userName
>   }})
> : null
> ```

@matheusgrandi you condition should be with `&&` and not `||`

**Comment by holuborhee** [comment]
There is already a discussion on this issue here https://github.com/typeorm/typeorm/issues/9316, this can be closed to favor that.
